packages/flutter/lib/src/rendering/sliver.dart

```diff
@@ -1742,46 +1440,599 @@ class RenderViewport2 extends RenderBox with ContainerRenderObjectMixin<RenderSl
+// ///   placed inside a [RenderSliver] (the opposite of this class).
+class RenderShrinkWrappingViewport extends RenderViewportBase2<SliverLogicalContainerParentData> {
+  /// Creates a viewport (for [RenderSliver] objects) that shrink-wraps its
+  /// contents.
+  ///
+  /// The [offset] must be specified. For testing purposes, consider passing a
+  /// [new ViewportOffset.zero] or [new ViewportOffset.fixed].
+  RenderShrinkWrappingViewport({
+    AxisDirection axisDirection: AxisDirection.down,
+    @required ViewportOffset offset,
+    List<RenderSliver> children,
+  }) : super(axisDirection: axisDirection, offset: offset) {
+    addAll(children);
+  }
+
+  @override
+  void setupParentData(RenderObject child) {
+    if (child.parentData is! SliverLogicalContainerParentData)
+      child.parentData = new SliverLogicalContainerParentData();
+  }
+
+  // Out-of-band data computed during layout.
+  double _maxScrollExtent;
+  double _shrinkWrapExtent;
+  bool _hasVisualOverflow = false;
+
+  @override
+  void performLayout() {
+    if (firstChild == null) {
+      switch (axis) {
+        case Axis.vertical:
+          size = constraints.constrain(new Size(constraints.maxWidth, 0.0));
+          break;
+        case Axis.horizontal:
+          size = constraints.constrain(new Size(0.0, constraints.maxHeight));
+          break;
+      }
+      offset.applyViewportDimension(0.0);
+      _maxScrollExtent = 0.0;
+      _shrinkWrapExtent = 0.0;
+      _hasVisualOverflow = false;
+      offset.applyContentDimensions(0.0, 0.0);
+      return;
+    }
+
+    double extent;
+    double crossExtent;
+    switch (axis) {
+      case Axis.vertical:
+        assert(constraints.hasBoundedWidth);
+        extent = constraints.maxHeight;
+        crossExtent = constraints.maxWidth;
+        break;
+      case Axis.horizontal:
+        assert(constraints.hasBoundedHeight);
+        extent = constraints.maxWidth;
+        crossExtent = constraints.maxHeight;
+        break;
+    }
+
+    double correction;
+    double effectiveExtent;
+    do {
+      assert(offset.pixels != null);
+      correction = _attemptLayout(extent, crossExtent, offset.pixels);
+      if (correction != 0.0) {
+        offset.correctBy(correction);
+      } else {
+        switch (axis) {
+          case Axis.vertical:
+            effectiveExtent = constraints.constrainHeight(_shrinkWrapExtent);
+            break;
+          case Axis.horizontal:
+            effectiveExtent = constraints.constrainWidth(_shrinkWrapExtent);
+            break;
+        }
+        offset.applyViewportDimension(effectiveExtent);
+        // when updating this, also update similar code in rereportDimensions
+        if (offset.applyContentDimensions(0.0, math.max(0.0, _maxScrollExtent - effectiveExtent)))
+          break;
+      }
+    } while (true);
+    switch (axis) {
+      case Axis.vertical:
+        size = constraints.constrain(new Size(crossExtent, effectiveExtent));
```

Comment:
I'm surprised that we don't have a version of `constrain` that takes the width and height separately.  Might be worth adding to avoid the extra `Size` allocation.